---
title: "Intron retention analysis"
author: "Xiaoli Mo"
institute: Xiamen University
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# 1. Differentially gene expression analysis

```{r differential expression analysis}
library(dplyr)
library(edgeR) 
library(ngscmdr)
library(edgeR)
library(scales)
library(ggplot2)

WT_counts <- read.table("WT.txt", header = T)
KO_counts <- read.table("Rbpm1_KO_counts_output.txt", header =T)

WT_counts <- WT_counts[,-2:9]
WT_counts <- WT_counts[,-2:-9]
View(KO_counts)
KO_counts <- KO_counts[,-1:-6]
WT_vs_KO_counts <- cbind(WT_counts,KO_counts)

rm_list <- read.table("rm_list.txt",header=T)

WT_KO_counts_2 <- filter(WT_vs_KO_counts, !row.names(WT_vs_KO_counts) %in% rm_list[,1])

dgelist_2 <- DGEList(counts = WT_KO_counts_2, group = group)
keep <- rowSums(cpm(dgelist_2) > 1 ) >= 2
dgelist_keep <- dgelist_2[keep, , keep.lib.sizes = FALSE]
dgelist_keep_norm <- calcNormFactors(dgelist_keep, method = 'TMM')
design <- model.matrix(~group)

dge_k <- estimateDisp(dgelist_keep_norm, design, robust = TRUE)
fit_k <- glmQLFit(dge_k, design, robust = TRUE)
lrt_K <- topTags(glmQLFTest(fit_k), n = nrow(dgelist_2$counts))
write.csv(lrt_K,"lrt_k2.csv")

edgeR_KO_WT <- read.csv("lrt_k2.csv",header=T) 
edgeR_KO_WT$logFC <- edgeR_KO_WT$logFC * -1

label_list <- read.csv("label_list.txt", sep="")

vol_data <- data.frame(gene = edgeR_KO_WT$X,
                       FDR = -log10(edgeR_KO_WT$FDR), 
                       lfc = edgeR_KO_WT$logFC)

vol_data <- mutate(vol_data, color = case_when(vol_data$lfc > 2 & vol_data$FDR > 1.3  ~ "Increased",
                                       vol_data$lfc < -2 & vol_data$FDR > 1.3  ~ "Decreased",
                                       ###vol_data$lfc > 2 & 
                                       vol_data$FDR < 1.3  ~ "Nonsignificant",
                                       #vol_data$lfc < -2 & vol_data$FDR < 1.3  ~ "Nonsignificant",
                                       vol_data$lfc > -2 & vol_data$lfc < 2 & vol_data$FDR > 1.3  ~ "Nonsignificant",
                                      ))

vol_data$color[vol_data$gene == "PY17X_1109100"] <- "Retained"
vol_data$color[vol_data$gene == "PY17X_0833600"] <- "Retained"
vol_data$color[vol_data$gene == "PY17X_0716700"] <- "Retained"
vol_data$color[vol_data$gene == "PY17X_1216400"] <- "Retained"

vol <- ggplot(vol_data %>%
              arrange(color), aes(x = lfc, y = FDR, color = color))

vol +    
    ggtitle(label = "Volcano Plot for Rbmp1 KO", subtitle = "Colored by fold-change direction") +
    geom_point(size = 2.5, alpha = 0.8, na.rm = T) +
    scale_color_manual(name = "Directionality",
                       values = c(Increased = "#008B00", Decreased = "#CD4F39", Nonsignificant = "darkgray")) +
    theme_bw(base_size = 14) + 
    theme(legend.position = "right") + 
    xlab(expression(log[2]("Fold Change (KO / WT)"))) +
    ylab(expression(-log[10]("FDR"))) + 
    geom_hline(yintercept = 1.3, colour = "darkgrey", linetype="dashed") + 
    geom_vline(xintercept = 2, colour = "darkgrey", linetype="dashed") + 
    geom_vline(xintercept = -2, colour = "darkgrey", linetype="dashed") +
    geom_text_repel(aes(label = gene),
                    data = vol_data[vol_data$gene  %in% label_list$gene_ID,], size=4, 
                    color = "black", min.segment.length = unit(0.1, 'lines'), 
                    nudge_y = 0.2) +
    geom_text_repel(aes(label ="Rbpm1"),  data = vol_data[vol_data$gene == "PY17X_0716700",],
                    size=4, color = "black", min.segment.length = unit(0.1, 'lines'), 
                    nudge_y = 0.2) +
    geom_point(aes(label = gene ),
               data = vol_data[vol_data$gene  %in% label_list$gene_ID,], color="#8B1A1A") +
    geom_point(aes(label = gene)m,
               data = vol_data[vol_data$gene == "PY17X_0716700",], color="#8B1A1A") +
    theme_bw() + 
    theme(panel.grid=element_blank()) + 
    scale_x_continuous(breaks= c(-10,-5,-2,0,2,5,10)) + facet_zoom(xlim= c(-5,-2), ylim =c(9,10.5) )

vol +    
    ggtitle(label = "Volcano Plot for Rbmp1 KO", subtitle = "Colored by fold-change direction") +
    geom_point(size = 2.5, alpha = 0.8, na.rm = T) +
    scale_color_manual(name = "Directionality",
                       values = c(Increased = "#008B00", Decreased = "#CD4F39", Nonsignificant = "darkgray")) +
    theme_bw(base_size = 14) + 
    theme(legend.position = "right") + 
    xlab(expression(log[2]("Fold Change (KO / WT)"))) +
    ylab(expression(-log[10]("FDR"))) + 
    geom_hline(yintercept = 1.3, colour = "darkgrey", linetype="dashed") + 
    geom_vline(xintercept = 2, colour = "darkgrey", linetype="dashed") + 
    geom_vline(xintercept = -2, colour = "darkgrey", linetype="dashed") +
    geom_text_repel(aes(label = gene),force = 20,segment.size=0.5,direction = "both",
                    data = vol_data[vol_data$gene  %in% label_list$gene_ID,], size=4, 
                    color = "black", min.segment.length = unit(0.1, 'lines'), 
                    nudge_y = 1.2) +
    geom_text_repel(aes(label ="Rbpm1"),  data = vol_data[vol_data$gene == "PY17X_0716700",],
                    size=4, color = "black", min.segment.length = unit(0.1, 'lines'), 
                    nudge_y = 0.6) +
    geom_point(
               data = vol_data[vol_data$gene  %in% label_list$gene_ID,], color="#8B1A1A") +
    geom_point(
               data = vol_data[vol_data$gene == "PY17X_0716700",], color="#8B1A1A") +
    theme_bw() + 
    theme(panel.grid=element_blank()) + 
    scale_x_continuous(breaks= c(-10,-5,-2,0,2,5,10)) + facet_zoom(xlim= c(-5,-1), ylim =c(2.5,12) )

vol +    
    ggtitle(label = "Volcano Plot for Rbmp1 KO", subtitle = "Colored by fold-change direction") +
    geom_point(size = 0.5, na.rm = T) +
    scale_color_manual(name = "Directionality",
                       values = c(Increased = "#C85F5F", Decreased = "#50B4FA", Nonsignificant = "#D7D7D7", Retained = "#1003DC")) +
    theme_bw(base_size = 14) + 
    theme(legend.position = "right") + 
    xlab(expression(log[2]("FC Mutant/Parental"))) +
    ylab(expression(-log[10]("FDR"))) + 
    geom_hline(yintercept = 1.3, colour = "darkgrey", linetype="dashed") + 
    geom_vline(xintercept = 2, colour = "darkgrey", linetype="dashed") + 
    geom_vline(xintercept = -2, colour = "darkgrey", linetype="dashed") +
    geom_text_repel(aes(label = gene),force = 50,segment.size=0.5,direction = "both",
                    data = vol_data[vol_data$gene  %in% label_list$gene_ID,], size=4, 
                    color = "black", min.segment.length = unit(0.1, 'lines'), 
                    nudge_y = 2.0) +
    geom_text_repel(aes(label ="Rbpm1"),  data = vol_data[vol_data$gene == "PY17X_0716700",],
                    size=4, color = "black", min.segment.length = unit(0.1, 'lines'), 
                    nudge_y = 0.6) +
    geom_point(
               data = vol_data[vol_data$gene  %in% label_list$gene_ID,],  size = 1.0) +
    geom_point(
               data = vol_data[vol_data$gene == "PY17X_0716700",], size = 1.0) +
    geom_point(
               data = vol_data[vol_data$gene  %in% label_list$gene_ID,], stroke = 0.3, 
               colour = "black", size = 1.3, shape = 21) +
    geom_point(
               data = vol_data[vol_data$gene == "PY17X_0716700",], stroke = 0.3,  
               colour = "black", size = 1.3, shape = 21) +
    theme_bw() + 
    theme(panel.grid=element_blank()) + 
    scale_x_continuous(breaks= seq(-14,14,2)) + 
    scale_y_continuous (breaks = seq(0,15,3))
        
KO_WT_logFC2_fdr_0.05 <- filter(edgeR_KO_WT, logFC >= 2, FDR <= 0.05)
write.csv(KO_WT_up_logFC2_fdr_0.05 ,"KO_WT_up_logFC2_fdr_0.05.csv")

KO_WT_logFC2_fdr_0.05 <- filter(edgeR_KO_WT, logFC <= -2, FDR <= 0.05)
write.csv(KO_WT_down_logFC2_fdr_0.05 ,"KO_WT_down_logFC2_fdr_0.05.csv")

KO_WT_logFC2_fdr_0.01 <- filter(edgeR_KO_WT, logFC >= 2, FDR <= 0.01)
write.csv(KO_WT_up_logFC2_fdr_0.01 ,"KO_WT_up_logFC2_fdr_0.01.csv")

KO_WT_logFC2_fdr_0.01 <- filter(edgeR_KO_WT, logFC <= -2, FDR <= 0.01)
write.csv(KO_WT_down_logFC2_fdr_0.01 ,"KO_WT_down_logFC2_fdr_0.01.csv")
```

# 2. Differentially intron retention analysis

```{r intron retetion analysis}
i_counts <- read.table("gene_w_intron_final_counts.tab",header=T) 
gene_w_intron <- read.delim("gene_w_intron.bed", header=FALSE) 
gene_w_intron <- gene_w_intron[,c(-5,-6,-7,-9)]
colnames(gene_w_intron) <- c("chr","start","end","geneID","type","intron_info")
id <- arrange(gene_w_intron,chr,start,end)
counts_id <- cbind(id,i_counts[,4:9])
IE_data <- filter(counts_id, type == "exon" | type == "intron")

intron_exon_signal <- data.frame(chr = character(0), start = numeric(0), end= numeric(0),
                                 geneID = character(0), type = character(0),intron_info = character(0), 
                                 KO_IE_signal_1_U = numeric(0), KO_IE_signal_1_D = numeric(0),
                                 KO_IE_signal_2_U = numeric(0), KO_IE_signal_2_D = numeric(0),  
                                 KO_IE_signal_3_U = numeric(0), KO_IE_signal_3_D = numeric(0),
                                 WT_IE_signal_1_U = numeric(0), WT_IE_signal_1_D = numeric(0), 
                                 WT_IE_signal_2_U = numeric(0), WT_IE_signal_2_D = numeric(0), 
                                 WT_IE_signal_3_U = numeric(0), WT_IE_signal_3_D = numeric(0)) 
for (i in 1: nrow(IE_data)) {
     if (isTRUE (IE_data[i,5] == "intron")) {
         temp1 <- rbind(IE_data[i-1,], IE_data[i,],IE_data[i+1,])
         temp2 <- IE_data[i,]
         temp2 <- mutate(temp2, 
                         KO_IE_signal_1_U = temp1[2,7] / temp1[1,7], KO_IE_signal_1_D = temp1[2,7] / temp1[3,7],
                         KO_IE_signal_2_U = temp1[2,8] / temp1[1,8], KO_IE_signal_2_D = temp1[2,8] / temp1[3,8],
                         KO_IE_signal_3_U = temp1[2,9] / temp1[1,9], KO_IE_signal_3_D = temp1[2,9] / temp1[3,9],
                         WT_IE_signal_1_U = temp1[2,10] / temp1[1,10], WT_IE_signal_1_D = temp1[2,10] / temp1[3,10],
                         WT_IE_signal_2_U = temp1[2,11] / temp1[1,11], WT_IE_signal_2_D = temp1[2,11] / temp1[3,11],
                         WT_IE_signal_3_U = temp1[2,12] / temp1[1,12], WT_IE_signal_3_D = temp1[2,12] / temp1[3,12])
         intron_exon_signal <- rbind(intron_exon_signal,temp2[,c(1:6, 13:24)])
      }
} 

intron_exon_signal[mapply(is.infinite, intron_exon_signal)] <- 300
intron_exon_signal[mapply(is.na, intron_exon_signal)] <- 0

keep <- rowSums(intron_exon_signal[,c(7,9,11)] >=0.5 ) >= 2
KO_U <- cbind(intron_exon_signal,keep)
KO_U <- filter(KO_U, keep == "TRUE")

keep <- rowSums(intron_exon_signal[,c(8,10,12)] >=0.5 ) >= 2
KO_D <- cbind(intron_exon_signal,keep)
KO_D <- filter(KO_D, keep == "TRUE")

KO <- rbind(KO_D,KO_U)
KO <- distinct(KO)

KO <- KO[,-19]

keep <- rowSums(intron_exon_signal[,c(13,15,17)] >=0.5 ) >= 2
WT_U <- cbind(intron_exon_signal,keep)
WT_U <- filter(WT_U, keep == "FALSE")

keep <- rowSums(intron_exon_signal[,c(14,16,18)] >=0.5 ) >= 2
WT_D <- cbind(intron_exon_signal,keep)
WT_D <- filter(WT_D, keep == "FALSE")

WT <- rbind(WT_D, WT_U)
WT <- distinct(WT)
WT <- WT[,-19]


signal_filter <- intersect(WT,KO)

KO_read <- read.table("Rbpm1_KO_counts_output.txt",header=T,skip=1)
View(KO_read)
WT_read <- read.table("2018_vs_2023_final_counts_output.txt",header=T,skip=1)

KO_TPM <- calc_tpm_from_featurecounts(KO_read)
WT_TPM <- calc_tpm_from_featurecounts(WT_read)
WT_male_TPM <- WT_TPM[,-2:-15]
KO_TPM <- KO_TPM[,-2:-9]

colnames(KO_TPM) <- gsub("X.cluster.facility.jyuan.ximo.Rbpm1.Py_2023.Rbpm1_KO.map.","",colnames(KO_TPM),fix=T)
colnames(KO_TPM) <- gsub(".bam.sort.bam","",colnames(KO_TPM),fix=T)

colnames(WT_male_TPM) <- gsub("X.cluster.facility.jyuan.ximo.Rbpm1.Py_2023.2018vs2023.map.","",colnames(WT_male_TPM),fix=T)
colnames(WT_male_TPM) <- gsub(".bam.sort.bam","",colnames(WT_male_TPM),fix=T)

WT_male_TPM_keep <- cbind(WT_male_TPM,rowMeans(WT_male_TPM[,2:4]) >= 30)
colnames(WT_male_TPM_keep) <- gsub("rowMeans(WT_male_TPM[, 2:4]) >= 30","factor", colnames(WT_male_TPM_keep),fix=T)
WT_male_f <- filter(WT_male_TPM_keep, factor == "TRUE")

KO_TPM_keep <- cbind(KO_TPM,rowMeans(KO_TPM[,2:4]) >= 30)
colnames(KO_TPM_keep) <- gsub("rowMeans(KO_TPM[, 2:4]) >= 30","factor", colnames(KO_TPM_keep),fix=T)
KO_male_f <- filter(KO_TPM_keep, factor == "TRUE")

all_f <- intersect(WT_male_f[,1],KO_male_f[,1])

all_f <- matrix(all_f)
colnames(all_f)[1] <- "GeneID"


WT_counts <-read.table("male_vs_female_final_counts_output",header = T)
WT_counts <- WT_counts[,-2:-9]  
colnames(WT_counts) <- gsub("X.cluster.facility.jyuan.ximo.Rbpm1.Py_2023.2018vs2023.map.","",
                            colnames(WT_counts),fix=T)
colnames(WT_counts) <- gsub(".bam.sort.bam","",colnames(WT_counts),fix = T)

KO_counts <-read.table("Rbpm1_KO_counts_output.txt",header = T)
KO_counts <- KO_counts[,-2:-6]
colnames(KO_counts) <- gsub("X.cluster.facility.jyuan.ximo.Rbpm1.Py_2023.Rbpm1_KO.map.Rbpm1_",""
                            ,colnames(KO_counts),fix=T) 
colnames(KO_counts) <- gsub(".bam.sort.bam","",colnames(KO_counts),fix = T)

gene_counts <- cbind(WT_counts,KO_counts[,2:4])

intron_counts <- read.table("intron_count_id.tab",header = T)
intron_counts <- select(intron_counts,1,2,3,4,5,6,10,11,12,7,8,9)

intron_counts_keep <- intron_counts[0,]

 for (i in 1:nrow(all_f)){
      temp1 <- all_f[i,]
      temp2 <- filter(intron_counts, geneID == temp1[1])
      intron_counts_keep <- rbind(intron_counts_keep,temp2)
 }


intron_counts_K <- cbind(intron_counts_keep[,1:6],intron_counts_keep[,7:12]*1000)

colnames(intron_counts_K) <- gsub("WT_","intron_WT_",colnames(intron_counts_K),fix = T)
colnames(intron_counts_K) <- gsub("KO_","intron_KO_",colnames(intron_counts_K),fix = T)

temp_op2 <- data.frame(intron_id=numeric(0),chr=character(0),start=numeric(0), end=numeric(0), GeneID=character(0),intronID=character(0),
                       intron_KO_rep_1_n=numeric(0),intron_KO_rep_2_n=numeric(0),intron_KO_rep_3_n=numeric(0),
                       intron_WT_rep_1_n=numeric(0),intron_WT_rep_2_n=numeric(0),intron_WT_rep_3_n=numeric(0)) 

for (i in 1:nrow(intron_counts_K)){
    
    temp1 <- intron_counts_K[i,]
    temp2 <- filter(gene_counts,Geneid == temp1[1,5])
    temp1 <- cbind(temp1, temp2[2:7]) 
      
        for(j in 7:9) {
            if (isTRUE(temp1[1,j+6] == 0)) {
                temp1 <- mutate(temp1, "intron_WT_rep_{j-6}_n" := 0)
                } else { temp1 <- mutate(temp1, "intron_WT_rep_{j-6}_n" := temp1[1,j]/temp1[1,j+6]) 
                }
                } 
   
                for(k in 10:12) {
            if (isTRUE(temp1[1,k+6] == 0)) {
                temp1 <- mutate(temp1, "intron_KO_rep_{k-9}_n" := 0)
                } else { temp1 <- mutate(temp1, "intron_KO_rep_{k-9}_n" := temp1[1,k]/temp1[1,k+6])
                }
                } 

    temp_op2 <- rbind (temp_op2,temp1[1,c(1:6,19:24)]) 
}  

RI_signal <- temp_op2

RI_signal_f <- RI_signal[0,]
for (i in 1:nrow(signal_filter)) {
    temp1 <- filter(RI_signal, intronID == signal_filter[i,6])
    RI_signal_f <- rbind (RI_signal_f, temp1)
}

n_counts <- RI_signal_f[,7:12]
group <- rep(c('WT', 'KO'), each = 3)

dinlist <- DGEList(counts =n_counts, group = group)
dinlist_norm <- calcNormFactors(dinlist, method = 'TMM')
design <- model.matrix(~group)


din <- estimateDisp(dinlist_norm, design, robust = TRUE)
fit <- glmQLFit(din, design, robust = TRUE)
lrt <- topTags(glmQLFTest(fit), n = nrow(dinlist$counts))
write.csv(lrt,"lrt.csv")

edgeR_intron <- read.csv("lrt.csv",header=T)
colnames(edgeR_intron)[1] <- "intron_id"

id_21 <- read.table("21_id.txt",header=T)

intron_ID <- read.table("intron_count_id.tab", header = T)
intron_ID <- intron_ID[,1:6]

edgeR_w_id <- data.frame(intron_id=numeric(0),chr=character(0),start=numeric(0), 
                         end=numeric(0), GeneID=character(0),intronID=character(0),
                         logFCd=numeric(0),logCPMd=numeric(0),F=numeric(0),PValue=numeric(0),
                         FDR=numeric(0))

for (i in 1:nrow(edgeR_intron)){
     temp1 <- edgeR_intron[i,]
     temp2 <- filter(intron_ID, intron_id == temp1[1,1])
     temp1 <- cbind(temp2, temp1[,2:6]) 
     edgeR_w_id <- rbind(edgeR_w_id,temp1)
}

edgeR_w_id$logFC <- edgeR_w_id$logFC * -1

vol_data <- data.frame(intron = edgeR_w_id$intron_id, 
                       gene = edgeR_w_id$geneID,
                       FDR = -log10(edgeR_w_id$FDR), 
                       lfc = edgeR_w_id$logFC)

vol_data <- mutate(vol_data, color = case_when(vol_data$lfc > 1 & vol_data$FDR > 1.3  ~ "Increased",
                                       vol_data$lfc < -1 & vol_data$FDR > 1.3  ~ "Decreased",
                                       vol_data$FDR < 1  ~ "nonsignificant",
                                       vol_data$lfc <= 1 & vol_data$lfc >= -1 ~ "nonsignificant"
                                      ))

vol <- ggplot(vol_data, aes(x = lfc, y = FDR, color = color))

squash_axis <- function(from, to, factor) { 
  trans <- function(x) {    
      isq <- x > from & x < to
      ito <- x >= to
      x[isq] <- from + (x[isq] - from)/factor
      x[ito] <- from + (to - from)/factor + (x[ito] - to)
      return(x)
  }

  inv <- function(x) {
      isq <- x > from & x < from + (to - from)/factor
      ito <- x >= from + (to - from)/factor
      x[isq] <- from + (x[isq] - from) * factor
      x[ito] <- to + (x[ito] - (from + (to - from)/factor))
      return(x)
  }

  return(trans_new("squash_axis", trans, inv))
}

 vol +   
     ggtitle(label = "Volcano Plot for Intron Rentetion Events", subtitle = "Colored by fold-change direction") +
             geom_point(size = 2.5, alpha = 0.8, na.rm = T) +
             scale_color_manual(name = "Directionality",
                                values = c(Increased = "#7CFC00", Decreased = "#CD4F39", nonsignificant = "darkgray")) +
                                theme_bw(base_size = 14) + 
             theme(legend.position = "right") + 
             xlab(expression(log[2]("Fold Change (KO / WT)"))) +
             ylab(expression(-log[10]("FDR"))) + 
             geom_hline(yintercept = 1.3, colour = "darkgrey", linetype = "dash") + 
             coord_trans(y = squash_axis(0, 1.3, 5)) + 
             geom_text(aes(label = gene), 
                       data = vol_data[vol_data$intron  %in% id_21$intron_id,],size=4, 
                       vjust=-1, hjust = 0.5, show.legend = FALSE,color="#008B00") +
             geom_point(data = vol_data[vol_data$intron  %in% id_21$intron_id,],color="#008B00")

write.csv(edgeR_w_id,"edgeR_all_intron.csv")

edgeR_logFC1_fdr_0.05 <- filter(edgeR_w_id, logFC >= 1, FDR <= 0.05)
write.csv(edgeR_logFC1_fdr_0.05 ,"edgeR_logFC1_fdr_0.05.csv")
```